tools/clang/include/clang/SPIRV/Constant.h

```diff
@@ -0,0 +1,116 @@
+/// constant. type. It also includes a set of decorations that are applied to
+/// that constant.
+///
+/// The class includes static getXXX(...) functions for getting pointers of any
+/// needed constant. A unique constant has a unique pointer (e.g. calling
+/// 'getTrue' function will always return the same pointer for the given
+/// context).
+class Constant {
+public:
+  using DecorationSet = std::set<const Decoration *>;
+
+  spv::Op getOpcode() const { return opcode; }
+  uint32_t getTypeId() const { return typeId; }
+  const std::vector<uint32_t> &getArgs() const { return args; }
+  const DecorationSet &getDecorations() const { return decorations; }
+  bool hasDecoration(const Decoration *) const;
+
+  // OpConstantTrue and OpConstantFalse are boolean
+  // OpSpecConstantTrue and OpSpecConstantFalse are boolean
+  bool isBoolean() const;
+
+  // OpConstant and OpSpecConstant are only allowed to take integers and floats
+  bool isNumerical() const;
+
+  // OpConstantComposite and OpSpecConstantComposite
+  bool isComposite() const;
+
+  // Get constants
+  static const Constant *getTrue(SPIRVContext &ctx, uint32_t type_id,
+                                 DecorationSet dec = {});
+  static const Constant *getFalse(SPIRVContext &ctx, uint32_t type_id,
+                                  DecorationSet dec = {});
+  static const Constant *getNumeric(SPIRVContext &ctx, uint32_t type_id,
+                                    llvm::ArrayRef<uint32_t> value,
+                                    DecorationSet dec = {});
+  static const Constant *getComposite(SPIRVContext &ctx, uint32_t type_id,
+                                      llvm::ArrayRef<uint32_t> constituents,
+                                      DecorationSet dec = {});
+  static const Constant *getSampler(SPIRVContext &ctx, uint32_t type_id,
+                                    spv::SamplerAddressingMode, uint32_t param,
+                                    spv::SamplerFilterMode,
+                                    DecorationSet dec = {});
+  static const Constant *getNull(SPIRVContext &ctx, uint32_t type_id,
+                                 DecorationSet dec = {});
+
+  // Get specialization constants
+  static const Constant *getSpecTrue(SPIRVContext &ctx, uint32_t type_id,
+                                     DecorationSet dec = {});
+  static const Constant *getSpecFalse(SPIRVContext &ctx, uint32_t type_id,
+                                      DecorationSet dec = {});
+  static const Constant *getSpecNumeric(SPIRVContext &ctx, uint32_t type_id,
+                                        llvm::ArrayRef<uint32_t> value,
+                                        DecorationSet dec = {});
+  static const Constant *getSpecComposite(SPIRVContext &ctx, uint32_t type_id,
+                                          llvm::ArrayRef<uint32_t> constituents,
+                                          DecorationSet dec = {});
+  // TODO: Implement OpSpecConstantOp if proven it adds any value
```

Comment:
No need for this TODO. It's more like implement something in the next step, which we will surely do when time comes.